File: libcrocofixpython/crocofix/sockets/scheduler.py
from .._crocofix import _Scheduler

import sched, time

class Scheduler(_Scheduler):

    def __init__(self):
        super().__init__()
        self.scheduler = sched.scheduler(time.time, time.sleep)

    def run(self):
        while True:
            self.scheduler.run()
            time.sleep(0.2)

    def schedule(self, task):
        # This is used by test code in the C++ implementation so we don't need it here.
        raise NotImplementedError()
 
    def schedule_relative_callback(self, when, callback):
        print("SCHEDULE RELATIVE {}".format(when.total_seconds()))
        # self.scheduler.enter(when.total_seconds(), priority=0, action=callback)
        return 0
    
    def schedule_repeating_callback(self, interval, callback):
        print("SCHEDULE REPEATING {}".format(interval))
        # task = asyncio.create_task(self.repeating_callback(interval, callback))
        # # task = asyncio.create_task(task_coroutine(), name="MyTask")
        # # asyncio.tasks.all_tasks. // This is a set
        # self.tasks.add(task)
        # task.add_done_callback(self.tasks.discard)
        # TODO
        return 0
 
    def cancel_callback(self, token):
        # TODO - we need this for session shutdown
        raise NotImplementedError()

File: libcrocofixpython/scheduler.cpp
#include "scheduler.hpp"
#include <libcrocofix/scheduler.hpp>
#include <pybind11/pybind11.h>
#include <pybind11/functional.h>
#include <pybind11/chrono.h>

#include <iostream>

namespace py = pybind11;

class Scheduler : public crocofix::scheduler
{
public:

    void run() override
    {
        PYBIND11_OVERRIDE_PURE(void, crocofix::scheduler, run,);
    }

    void schedule(task_type task) override
    {
        PYBIND11_OVERRIDE_PURE(void, crocofix::scheduler, schedule, task);
    }

    cancellation_token schedule_relative_callback(std::chrono::milliseconds when, const scheduled_callback& callback) override
    {
        PYBIND11_OVERRIDE_PURE(cancellation_token, crocofix::scheduler, schedule_relative_callback, when, callback);
    }

    cancellation_token schedule_repeating_callback(std::chrono::milliseconds interval, const scheduled_callback& callback) override
    {
        PYBIND11_OVERRIDE_PURE(cancellation_token, crocofix::scheduler, schedule_repeating_callback, interval, callback);
    }

    void cancel_callback(cancellation_token token) override
    {
        PYBIND11_OVERRIDE_PURE(void, crocofix::scheduler, cancel_callback, token);
    }
};

void init_scheduler(py::module_& module)
{
    py::class_<crocofix::scheduler, Scheduler>(module, "_Scheduler")
        .def(py::init<>())
        .def("run", &crocofix::scheduler::run)
        .def("schedule", &crocofix::scheduler::schedule)
        .def("schedule_relative_callback", &crocofix::scheduler::schedule_relative_callback)
        .def("schedule_repeating_callback", &crocofix::scheduler::schedule_repeating_callback)
        .def("cancel_callback", &crocofix::scheduler::cancel_callback)
    ;
}
